### PR TITLE
[GEP-15] Handle extensionsv1alpha1.Bastion resources

### PR DESCRIPTION
**Pull request by xrstf**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement
/platform aws

**What this PR does / why we need it**:
This adds support for the new Bastion resource, which is part of [GEP-15](https://github.com/gardener/gardener/blob/master/docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md)'s new SSH handling. A new controller takes care of creating a bastion host with the configured SSH key and allowing ingress from that host to the other worker nodes in a shoot.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Be gentle, it's my first time.

**Release note**:

```other operator
Handle extensionsv1alpha1.Bastion resources for SSH access to worker instances
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @xrstf for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by petersutter** [comment]
@xrstf can you restrict the security group of the bastion to the bare minimum? E.g. no egress traffic should be allowed and is not needed. Only the connection from the bastion to the shoot nodes should be required I presume.
Maybe we can add that also to the GEP-15

cc @ThormaehlenFred

**Comment by xrstf** [comment]
@petersutter I have implemented the additional egress restriction for the bastion's security group. The bastion security group can now only establish SSH connections to the worker security group.

File: pkg/controller/bastion/bastion_suite_test.go
Comment (rfranzke) [inline pkg/controller/bastion/bastion_suite_test.go:26]:
```suggestion
	RunSpecs(t, "Controller Bastion Suite")
```

File: pkg/controller/bastion/options.go
Comment (rfranzke) [inline pkg/controller/bastion/options.go:29]: Typically, we group the imports in 3 groups: 1) core, 2) own packages, 3) third-party packages

```suggestion
	awsextensionsv1alpha1 "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/v1alpha1"
	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/ec2"
	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
	"github.com/gardener/gardener/pkg/extensions"
	"github.com/pkg/errors"
```

Same for all other files

File: pkg/controller/bastion/actuator_delete.go
Comment (rfranzke) [inline pkg/controller/bastion/actuator_delete.go:35]: Can you rather pass the cluster as argument to the `Delete` (and `Reconcile`) functions? This is also how the `Actuator` interfaces of other extension resources look like: https://github.com/gardener/gardener/blob/c05c128cc126ccecb793cd5c72c72f93e8d0b5be/extensions/pkg/controller/controlplane/actuator.go#L26-L30

File: pkg/controller/bastion/actuator.go
Comment (rfranzke) [inline pkg/controller/bastion/actuator.go:57]: FYI: There is `github.com/gardener/gardener/pkg/utils/kubernetes.Key` which might be more convenient

File: pkg/controller/bastion/options.go
Comment (rfranzke) [inline pkg/controller/bastion/options.go:73]:
```suggestion
	cloudProfileConfig, err := getCloudProfileConfig(cluster)
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice PR!

Can you extend https://github.com/gardener/gardener-extension-provider-aws/blob/master/charts/gardener-extension-provider-aws/doc.go#L15 and run `make generate`? I assume you manually added the `Bastion/aws` type to the `example/controller-registration.yaml`.

Also, do you plan to have some test for the controller? I'd love to see some integration test using the envtest package, similar to what we do for the `Infrastructure` resource: https://github.com/gardener/gardener-extension-provider-aws/tree/master/test/integration/infrastructure

File: pkg/controller/bastion/actuator.go
Comment (xrstf) [reply pkg/controller/bastion/actuator.go:57]: I'm not sure the `Key()` function really offers anything of value here:

```
key := kubernetes.Key(v1beta1constants.SecretNameCloudProvider, bastion.Namespace)
// vs.
key := types.NamespacedName{Name: v1beta1constants.SecretNameCloudProvider, Namespace: bastion.Namespace}
```

Personally I would find the latter option to be more clear. I can change it if you insist though. ;)

File: pkg/controller/bastion/actuator_delete.go
Comment (xrstf) [reply pkg/controller/bastion/actuator_delete.go:56]: Yes, I've drawn a line _somewhere_ :-D I'm not sure we can defend against each and every user interaction. What if a user stops the bastion host? What if they add more security groups to the instance? What if they delete the SSH key from the bastion instance?

How deep do we want to go here? :)

File: pkg/controller/bastion/actuator_reconcile.go
Comment (xrstf) [reply pkg/controller/bastion/actuator_reconcile.go:80]: Indeed

File: pkg/controller/bastion/actuator_reconcile.go
Comment (xrstf) [reply pkg/controller/bastion/actuator_reconcile.go:69]: Because I forgot that I implemented a `Ready` in the first place.

File: pkg/controller/bastion/actuator_delete.go
Comment (petersutter) [reply pkg/controller/bastion/actuator_delete.go:56]: but why not get rid of this "early exit" and just let the rest of the cleanup logic run through?

Comment (xrstf) [reply pkg/controller/bastion/actuator_delete.go:52]: Because we need the SG ID for the worker-group cleanup. I can skip that if we don't have a security group anymore and just nuke the instance instead.

Comment (xrstf) [reply pkg/controller/bastion/actuator_delete.go]: Done that now.

File: pkg/controller/bastion/options.go
Comment (xrstf) [reply pkg/controller/bastion/options.go]: How consistent is this across the Gardener codebase? I see can lots of code that seems to follow different rules.

Comment (rfranzke) [reply pkg/controller/bastion/options.go]: There might be individual outliers, but we strive to keep it as consistent as possible...

File: example/20-crd-bastion.yaml
Comment (xrstf) [reply example/20-crd-bastion.yaml]: Done

**Comment by xrstf** [comment]
@petersutter @rfranzke I have added some integration tests, but have no idea if they will work in the CI environment as well. Running them locally via `go test -v . -access-key-id "..." -secret-access-key "...." -region eu-west-1` works fine, but I have my doubts regarding the AMI (the tests are setting a static AMI for the bastion host, which is maybe not available where the tests are running?).

File: pkg/controller/bastion/options.go
Comment (rfranzke) [inline pkg/controller/bastion/options.go:34]: doc string for exported struct missing

File: pkg/controller/bastion/options.go
Comment (rfranzke) [inline pkg/controller/bastion/options.go:22]:
```suggestion
	awsv1alpha1 "github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/v1alpha1"
```

File: test/integration/subnet.go
Comment (rfranzke) [inline test/integration/subnet.go:29]: doc string for exported functions missing in this file

File: test/integration/securitygroup.go
Comment (rfranzke) [inline test/integration/securitygroup.go:27]: doc string for exported functions missing in this file

File: test/integration/bastion/bastion_test.go
Comment (rfranzke) [inline test/integration/bastion/bastion_test.go:166]: Can this be removed?

**Review by rfranzke — COMMENTED** [review]
Thanks, from another high-level scan I only found a few minor nits. @petersutter PTAL

/hold - needs to revendor g/g once https://github.com/gardener/gardener/pull/3865 is merged
/cc @schrodit @dguendisch - anything we need to consider now that we will have two integration tests (see `.test-defs`)? Will `/test` execute both?

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [inline test/integration/bastion/bastion_test.go:62]: to be on the safe side we should only use private ip ranges for the test so that no one can access the test bastion

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [inline test/integration/bastion/bastion_test.go:412]: who has the private key? :D
As alternative we could also generate a new one for each test. WDYT?

Comment (xrstf) [reply test/integration/bastion/bastion_test.go:412]: It's the same key that is used in the infrastructure tests: https://github.com/gardener/gardener-extension-provider-aws/blob/master/test/integration/infrastructure/infrastructure_test.go#L424

File: test/integration/bastion/bastion_test.go
Comment (xrstf) [reply test/integration/bastion/bastion_test.go:166]: It can! Let me work my magic..

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [inline test/integration/bastion/bastion_test.go:103]: should the name also have an indicator that it's from a test in case it needs resources leak and need to manually be cleaned up?
@rfranzke how is this done for the other tests that create infrastructure resources? Is there already a naming convention or how is this handled?

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [inline test/integration/bastion/bastion_test.go:350]: @rfranzke how is this done for other tests that need to spin up a machine? are the values (version, ami) hardcoded or dynamically fetched

Comment (xrstf) [reply test/integration/bastion/bastion_test.go:350]: I presume we have to fetch it, as tests are probably not running in eu-west-1. I was unsure at first, so I made the simplest solution for now. ^^

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [reply test/integration/bastion/bastion_test.go:411]: seems like @ialidzhikov introduced this

@ialidzhikov do you see a (security) issue that this public key is used for the bastion test?

File: test/integration/subnet.go
Comment (petersutter) [inline test/integration/subnet.go:75]:
```suggestion
// DestroySubnet deletes an existing subnet.
```

Comment (xrstf) [reply test/integration/subnet.go]: Whoops, sorry.

File: test/integration/bastion/bastion_test.go
Comment (xrstf) [reply test/integration/bastion/bastion_test.go:350]: @rfranzke can you clarify your thumbs up emoji? ^^

Comment (rfranzke) [reply test/integration/bastion/bastion_test.go:350]: You already wrote it:

> I presume we have to fetch it, as tests are probably not running in eu-west-1.

As the region is configurable (and probably should be) the AMI needs to be fetched either dynamically or also configured by the test executor.

Comment (schrodit) [reply test/integration/bastion/bastion_test.go:350]: > how is this done for other tests that need to spin up a machine? are the values (version, ami) hardcoded or dynamically fetched

all other tests currently have a configuration where the test executor passes the needed values.
For the shoot integration tests we use the `testrunner` that fetches the latest k8s version from the cloudprofile and passes it to the test. (samme for the os version).

Comment (xrstf) [reply test/integration/bastion/bastion_test.go]: The AMI is now determined dynamically based on the AMI name. From my tests, it seems the `ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210223` AMI is available in all regions.

Comment (schrodit) [reply test/integration/bastion/bastion_test.go]: as the hardcoded image is only for the test and the ami is dynamically fetched based on the region this is fine for the test.
cc @dguendisch

File: test/integration/bastion/bastion_test.go
Comment (schrodit) [inline test/integration/bastion/bastion_test.go:365]: can we have that version also as a constant like the image name? so that there is one place for such configuration

**Review by schrodit — CHANGES_REQUESTED** [review]
The test should be also added to the default testrun to be executed: https://github.com/gardener/gardener-extension-provider-aws/blob/master/.ci/testruns/default/templates/testrun.yaml#L33

If the bastion test should not run with the default test, a new testrun template has to be created and the testrun added in the bot config: https://github.com/gardener/gardener-extension-provider-aws/blob/master/.ci/testruns/default/templates/testrun.yaml#L33

File: .test-defs/bastion-test.yaml
Comment (xrstf) [reply .test-defs/bastion-test.yaml]: I applied your suggestion, but am not sure how this improves the cleanup by Ginkgo. The bastion tests already create cleanup actions, I presume without a test timeout that's lower than the pod lifetime, those would not be called and that's why we need a `go test` timeout? If so, why does the infrastructure test not have a timeout?

**Comment by xrstf** [comment]
> If the bastion test should not run with the default test, a new testrun template has to be created and the testrun added in the bot config: https://github.com/gardener/gardener-extension-provider-aws/blob/master/.ci/testruns/default/templates/testrun.yaml#L33

Would there be a reason for them not to run by default? @schrodit The bastion tests are usually pretty quick (less than 2 minutes) and seem to be stable enough to run them by default.

**Comment by schrodit** [comment]
> > If the bastion test should not run with the default test, a new testrun template has to be created and the testrun added in the bot config: https://github.com/gardener/gardener-extension-provider-aws/blob/master/.ci/testruns/default/templates/testrun.yaml#L33
> 
> Would there be a reason for them not to run by default? @schrodit The bastion tests are usually pretty quick (less than 2 minutes) and seem to be stable enough to run them by default.

No I think no specific. Its rather if they are slow and you only want to run them if related code has changed.
But if they're only 2 minutes lets run them with the default test suite

File: .test-defs/bastion-test.yaml
Comment (schrodit) [reply .test-defs/bastion-test.yaml]: In the gardener integration tests we have a specific timeout for each test that why we do not need it there.
But your're right we missed that in the aws infrastructure test. Can you add it there as well?

Comment (xrstf) [reply .test-defs/bastion-test.yaml]: Can and did. :)

File: test/integration/bastion/bastion_test.go
Comment (rfranzke) [reply test/integration/bastion/bastion_test.go:108]: Maybe something like this could be done as well: https://github.com/gardener/gardener-extension-provider-aws/blob/0d0b8ab4cb62f40116208ca371d3141bbaaf599d/test/integration/infrastructure/infrastructure_test.go#L453-L460

Comment (xrstf) [reply test/integration/bastion/bastion_test.go]: Changed it to `aws-bastion-it-[randomstr]`.

File: test/integration/bastion/bastion_test.go
Comment (rfranzke) [reply test/integration/bastion/bastion_test.go:432]: Why should this be an issue? It's (1) a public key which is (2) only used in a test. Let's keep it.

**Comment by rfranzke** [comment]
```
> Generate / Vendor Check
>> make generate
make generate needs to be run:
 M example/controller-registration.yaml
```

Also, I guess you could now vendor a commit of the `master` branch of `github.com/gardener/gardener` to make this buildable, or?

File: test/integration/bastion/bastion_test.go
Comment (petersutter) [reply test/integration/bastion/bastion_test.go:432]: we have test-private-keys all over the place so I wanted to be sure that this private key is not stored anywhere

**Review by petersutter — APPROVED** [review]
/lgtm

please also fix the TBD release note

**Review by schrodit — APPROVED** [review]
looks good from test pov

**Comment by xrstf** [comment]
@petersutter @rfranzke regenerated/resorted/revendored, should now be good to go.

File: pkg/controller/bastion/actuator.go
Comment (rfranzke) [inline pkg/controller/bastion/actuator.go:41]: doc strings for exported variables missing

**Review by rfranzke — COMMENTED** [review]
/reviewed ok-to-test

**Comment by rfranzke** [comment]
lgtm and mergeable fmpov after https://github.com/gardener/gardener-extension-provider-aws/pull/318#pullrequestreview-650185401 was addressed :)

**Review by petersutter — APPROVED** [review]
/lgtm

**Comment by petersutter** [comment]
one thing I just noticed in the test logs. I see 29 occurences of `error="observed generation outdated (0/1)"`

> 2021-05-03T09:52:36.530Z	INFO	infrastructure-actuator.terraformer	Waiting for Terraformer pod to be completed, pod hasn't finished yet	{"infrastructure": "aws-infrastructure-it--es7vz/infrastructure", "operation": "reconcile", "command": "apply", "pod": "aws-infrastructure-it--es7vz/infrastructure.infra.tf-apply-psv7d", "phase": "Running", "len-of-containerstatuses": 1}
> 2021-05-03 11:52:35	
> time="2021-05-03T09:52:35Z" level=error msg="Infrastructure aws-infrastructure-it--es7vz/infrastructure did not get ready yet" error="observed generation outdated (0/1)"

**Comment by xrstf** [comment]
@petersutter Can you explain the importance of that warning message? I gather something is continuously updating the Infrastructure resource, making some other component fail in updating/patching it?

File: pkg/controller/bastion/actuator.go
Comment (xrstf) [reply pkg/controller/bastion/actuator.go]: Done.

**Comment by petersutter** [comment]
I'm not sure if this is an "expected" behaviour and just wanted to bring it up as I noticed it. @rfranzke is this normal?

**Comment by rfranzke** [comment]
I think this is how the "wait until reconciled" function works, yes. Anyhow, unrelated to this PR.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by petersutter — DISMISSED** [review]

**Comment by rfranzke** [comment]
/unhold as the release was made
Though,
/needs rebase

**Comment by xrstf** [comment]
Rebased.

**Review by rfranzke — APPROVED** [review]

**Pull request merged**
